### PR TITLE
Add pipeline including speaker target tensor extracted from rttm

### PR DESCRIPTION
**Pull request by kingformatty**
# What does this PR do ?
Adding pipeline of fast-conformer_RNNT with speaker target tensors

**Collection**: asr

# Changelog 
- (1) Dataloader: LhotseSpeechToTextSpkBpeDataset (general)
- (2) Model: EncDecRNNTSPKBPEModel
- (3) Training Script: speech_to_text_rnnt_spk_bpe.py

# Usage
* You can potentially add a usage example below

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/asr/data/audio_to_text_lhotse_speaker.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+                                compute_num_samples(rttm_sup.end, cut.sampling_rate)
+                                if rttm_sup.end < cut.duration
+                                else compute_num_samples(rttm_sup.duration, cut.sampling_rate)
+                            )                   
+                    mask[speaker_idx, int(np.ceil(np.ceil((st + 1) / num_sample_per_mel_frame) / num_mel_frame_per_asr_frame)):int(np.ceil(np.ceil((et + 1) / num_sample_per_mel_frame) / num_mel_frame_per_asr_frame))] = 1
+
+        return mask
+
+
+class TokenizerWrapper:
```
Comment (tango4j) [inline nemo/collections/asr/data/audio_to_text_lhotse_speaker.py]: What is the need for this `TokenizerWrapper` to be re-defined?
If there is no change from the original one, please import from the original.

File: nemo/collections/asr/data/audio_to_text_lhotse_speaker.py [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        speaker_ats = [s.speaker for s in segments if not (s.speaker in seen or seen_add(s.speaker))]
+        
+        speaker_to_idx_map = {
+                spk: idx
+                for idx, spk in enumerate(speaker_ats)
+        }
+
+        
+        #initialize mask matrices (num_speaker, encoder_hidden)
+        mask = np.zeros((num_speakers, int(np.ceil(np.ceil((cut.num_samples + 1) / num_sample_per_mel_frame) / num_mel_frame_per_asr_frame))))
```
Comment (tango4j) [inline nemo/collections/asr/data/audio_to_text_lhotse_speaker.py]: This is better to be a standalone static function: 

@staticmethod
def get_frame_length_from_sample_length(num_samples):
 ...

File: nemo/collections/asr/models/__init__.py [hunk 3]
```diff
@@ -40,6 +40,9 @@
 from nemo.collections.asr.models.label_models import EncDecSpeakerLabelModel
 from nemo.collections.asr.models.msdd_models import EncDecDiarLabelModel, NeuralDiarizer
 from nemo.collections.asr.models.rnnt_bpe_models import EncDecRNNTBPEModel
+
```
Comment (tango4j) [inline nemo/collections/asr/models/__init__.py]: please remove the unnecessary spaces

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 4]
```diff
@@ -0,0 +1,785 @@
...
+from nemo.collections.asr.parts.submodules.rnnt_decoding import RNNTBPEDecoding, RNNTBPEDecodingConfig
+from nemo.collections.asr.parts.utils.asr_batching import get_semi_sorted_batch_sampler
+from nemo.collections.common.data.lhotse import get_lhotse_dataloader_from_config
+from nemo.core.classes.common import PretrainedModelInfo
+from nemo.utils import logging, model_utils
+
+
+
+
+class EncDecRNNTSPKBPEModel(EncDecRNNTModel, ASRBPEMixin):
```
Comment (tango4j) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py]: I think it is better to be `EncDecRNNTSpkBPEModel` or `EncDecRNNTBPESpkModel`  Since Spk is not an acronym, it is abbreviation.
Also a series of all caps letters is hard to read and parse the meaning

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 5]
```diff
@@ -0,0 +1,785 @@
...
+    """Base class for encoder decoder RNNT-based models with subword tokenization."""
+
+    @classmethod
+    def list_available_models(cls) -> List[PretrainedModelInfo]:
```
Comment (tango4j) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py:49]: Erase all the models here since there is no available model now.

File: nemo/collections/asr/models/rnnt_spk_bpe_models.py [hunk 6]
```diff
@@ -0,0 +1,785 @@
...
+
+        # Setup fused Joint step if flag is set
+        if self.joint.fuse_loss_wer:
+            self.joint.set_loss(self.loss)
+            self.joint.set_wer(self.wer)
+
+        # Setup projection layer
+        self.enc_spk_proj = torch.nn.Linear(self.cfg.model_defaults.enc_hidden + self.cfg.model_defaults.num_speakers, self.cfg.model_defaults.enc_hidden)
+
+    def change_vocabulary(
```
Comment (tango4j) [inline nemo/collections/asr/models/rnnt_spk_bpe_models.py]: All these functions should not be duplicated. 
If the function is in `EncDecRNNTBPEModel`, it needs to be inherited.
Not just this function, try minimizing the duplicated function so we can keep the code base minimal.

**Review by tango4j — CHANGES_REQUESTED** [review]
I did a quick first pass. Please reflect the comments I wrote.

File: nemo/collections/asr/data/audio_to_text_lhotse_speaker.py [hunk 7]
```diff
@@ -112,8 +108,8 @@ def speaker_to_target(self, cut, num_speakers: int = 4, num_sample_per_mel_frame
         }
 
         
-        #initialize mask matrices (num_speaker, encoder_hidden)
-        mask = np.zeros((num_speakers, int(np.ceil(np.ceil((cut.num_samples + 1) / num_sample_per_mel_frame) / num_mel_frame_per_asr_frame))))
+        #initialize mask matrices (num_speaker, encoder_hidden_len)
+        mask = np.zeros((num_speakers, self.get_hidden_length_from_sample_length(cut.num_samples, num_sample_per_mel_frame, num_mel_frame_per_asr_frame)))
```
Comment (weiqingw4ng) [inline nemo/collections/asr/data/audio_to_text_lhotse_speaker.py]: Better to use a variable for `self.get_hidden_length_from_sample_length(cut.num_samples, num_sample_per_mel_frame, num_mel_frame_per_asr_frame)`, e.g., `encoder_hidden_len ` as in your comment.

File: nemo/collections/asr/data/audio_to_text_lhotse_speaker.py [hunk 8]
```diff
@@ -130,34 +126,10 @@ def speaker_to_target(self, cut, num_speakers: int = 4, num_sample_per_mel_frame
                                 if rttm_sup.end < cut.duration
                                 else compute_num_samples(rttm_sup.duration, cut.sampling_rate)
                             )                   
-                    mask[speaker_idx, int(np.ceil(np.ceil((st + 1) / num_sample_per_mel_frame) / num_mel_frame_per_asr_frame)):int(np.ceil(np.ceil((et + 1) / num_sample_per_mel_frame) / num_mel_frame_per_asr_frame))] = 1
+                    mask[speaker_idx, self.get_hidden_length_from_sample_length(st, num_sample_per_mel_frame, num_mel_frame_per_asr_frame):self.get_hidden_length_from_sample_length(et, num_sample_per_mel_frame, num_mel_frame_per_asr_frame)] = 1
```
Comment (weiqingw4ng) [inline nemo/collections/asr/data/audio_to_text_lhotse_speaker.py]: Same for this.

File: nemo/collections/asr/data/audio_to_text_lhotse_speaker.py [hunk 9]
```diff
@@ -91,7 +76,18 @@ def __getitem__(self, cuts) -> Tuple[torch.Tensor, ...]:
...
+
+        Args:
+            cut: An audio cut
+            num_speakers: max number of speakers for all cuts ("mask" dim0), 4 by default
+            num_sample_per_mel_frame: number of sample per mel frame, sample_rate / 1000 * window_stride, 160 by default (10ms window stride)
+            num_mel_frame_per_asr_frame: encoder subsampling_factor, 8 by default
+            spk_tar_all_zero: set to True gives all zero "mask"
+        
+        Returns:
+            mask: speaker mask with shape (num_speaker, hidden_lenght)
```
Comment (weiqingw4ng) [inline nemo/collections/asr/data/audio_to_text_lhotse_speaker.py]: Typo hidden_length.

**Review by weiqingw4ng — COMMENTED** [review]
Just some minor comments. The major problems have been resolved in Taejin's review.
